File: Solutions/938_Range_Sum_of_BST/938_Range_Sum_of_BST.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rangeSumBST(self, root: TreeNode, low: int, high: int) -> int:
        def dfs(node):
            if not node: return
            if low <= node.val <= high: self.out += node.val
            if node.val > low:  dfs(node.left)
            if node.val < high: dfs(node.right)
                
        self.out = 0
        dfs(root)
        return self.out

#Fourth solution
class Solution(object):
    def rangeSumBST(self, root, L, R):
        def dfs(node):
            if node:
                if L <= node.val <= R:
                    self.ans += node.val
                if L < node.val:
                    dfs(node.left)
                if node.val < R:
                    dfs(node.right)

        self.ans = 0
        dfs(root)
        return self.ans

#Third fastest
class Solution:
    def rangeSumBST(self, root: TreeNode, low: int, high: int) -> int:
        total = 0
        def recurse(root):
            nonlocal total
            if not root:
                return
            if low <= root.val <= high:
                total += root.val
            if low < root.val:
                recurse(root.left)
            if high > root.val:
                recurse(root.right)
        recurse(root)
        return total

#Second fastest
class Solution:
    def rangeSumBST(self, root: TreeNode, L: int, R: int) -> int:
        def dfs(node):
            if node:
                if L <= node.val <= R:
                    self.ans += node.val
                if node.val > L:
                    dfs(node.left)
                if node.val < R:
                    dfs(node.right)

        self.ans = 0
        dfs(root)
        return self.ans

#Fastest
class Solution:
    def rangeSumBST(self, root: TreeNode, L: int, R: int) -> int:
        self.ans = 0
        def dfs(head):
            if  head : 
                if L <= head.val <= R:
                    self.ans +=head.val 
                if L <head.val :
                    dfs(head.left)
                if head.val <R :
                    dfs(head.right)
        dfs(root)
        return self.ans

File: Solutions/938_Range_Sum_of_BST/938_Range_Sum_of_BST.rs
//Second fastest
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;
impl Solution {
    pub fn range_sum_bst(root: Option<Rc<RefCell<TreeNode>>>, l: i32, r: i32) -> i32 {
        if let Some(root) = root {
            let mut sum = 0;
            if l <= root.borrow().val && root.borrow().val <=r {
                sum += root.borrow().val;
            }
            
            if l < root.borrow().val {
                sum += Self::range_sum_bst(root.borrow().left.clone(), l, r);
            }
            
            if r > root.borrow().val {
                sum += Self::range_sum_bst(root.borrow().right.clone(), l, r);
            }
            return sum;
        }
        return 0;
    }
}

//Another
use std::rc::Rc;
use std::cell::RefCell;

impl Solution {
    pub fn range_sum_bst(root: Option<Rc<RefCell<TreeNode>>>, low: i32, high: i32) -> i32 {
        match root {
            None => { 0 }
            Some(node) => {
                let node1 = Rc::try_unwrap(node).unwrap().into_inner();
                let left = Solution::range_sum_bst(node1.left, low, high);
                let right = Solution::range_sum_bst(node1.right, low, high);
                let current = if (low <= node1.val) && (node1.val <= high) { node1.val} else {0};
                left + right + current
            }
        }
    }
}

//Fastest
use std::rc::Rc;
use std::cell::RefCell;
impl Solution {
    pub fn range_sum_bst(root: Option<Rc<RefCell<TreeNode>>>, l: i32, r: i32) -> i32 {
        if root.is_none() {
            return 0;
        }
        let node_rc = root.unwrap();
        let node = node_rc.borrow();
        if node.val < l {
            return Solution::range_sum_bst(node.right.clone(), l, r);
        } else if node.val > r {
            return Solution::range_sum_bst(node.left.clone(), l, r);
        }
        node.val + Solution::range_sum_bst(node.right.clone(), l, r) +  Solution::range_sum_bst(node.left.clone(), l, r)
    }
}